src/Domain/DiagnosticInfo.hpp

```diff
@@ -0,0 +1,109 @@
+std::string diagnostic_info(const Domain<Dim>& domain,
+                            const Parallel::GlobalCache<Metavariables>& cache,
+                            const std::vector<size_t>& elements_per_core,
+                            const std::vector<size_t>& elements_per_node,
+                            const std::vector<size_t>& grid_points_per_core,
+                            const std::vector<size_t>& grid_points_per_node) {
+  const size_t total_number_of_elements =
+      alg::accumulate(elements_per_node, 0_st);
+  const size_t total_number_of_grid_points =
+      alg::accumulate(grid_points_per_node, 0_st);
+  const size_t total_number_of_blocks = domain.blocks().size();
+  const size_t number_of_cores = Parallel::number_of_procs<size_t>(cache);
+  const size_t number_of_nodes = Parallel::number_of_nodes<size_t>(cache);
+
+  // Sanity checks
+  ASSERT(total_number_of_elements == alg::accumulate(elements_per_core, 0_st),
+         "Number of elements determined from elements_per_core ("
+             << alg::accumulate(elements_per_core, 0_st)
+             << ") does not match number of elements determined from "
+                "elements_per_node ("
+             << total_number_of_elements << ").");
+  ASSERT(total_number_of_grid_points ==
+             alg::accumulate(grid_points_per_core, 0_st),
+         "Number of grid points determined from grid_points_per_core ("
+             << alg::accumulate(grid_points_per_core, 0_st)
+             << ") does not match number of grid points determined from "
+                "grid points_per_node ("
+             << total_number_of_grid_points << ").");
+  ASSERT(number_of_cores == elements_per_core.size(),
+         "Number of cores determined from elements_per_core ("
+             << elements_per_core.size()
+             << ") does not match number of cores determined from "
+                "global cache ("
+             << number_of_cores << ").");
+  ASSERT(number_of_cores == grid_points_per_core.size(),
+         "Number of cores determined from grid_points_per_core ("
+             << grid_points_per_core.size()
+             << ") does not match number of cores determined from "
+                "global cache ("
+             << number_of_cores << ").");
+  ASSERT(number_of_nodes == elements_per_node.size(),
+         "Number of nodes determined from elements_per_node ("
+             << elements_per_node.size()
+             << ") does not match number of nodes determined from "
+                "global cache ("
+             << number_of_nodes << ").");
+  ASSERT(number_of_nodes == grid_points_per_node.size(),
+         "Number of nodes determined from grid_points_per_node ("
+             << grid_points_per_node.size()
+             << ") does not match number of nodes determined from "
+                "global cache ("
+             << number_of_nodes << ").");
+
+  using ::operator<<;
+  std::stringstream ss{};
+  ss << "----- Domain Info -----\n";
```

Comment:
why make all these separate lines? I think you can actually just do:
```cpp
return MakeString{} << all_your_stuff;
```
to avoid dealing with the stringstream manually. Otherwise no need for a new `ss <<` each line.

Reply:
Just removed the `ss <<` so the code formatting stayed nice